src/kademlia/node.cpp

```diff
@@ -1191,28 +1190,60 @@ void node::write_nodes_entries(sha1_hash const& info_hash
 		if (wanted_node == m_nodes.end()) continue;
 		std::vector<node_entry> n;
 		wanted_node->second->m_table.find_node(info_hash, n, 0);
-		r[wanted_node->second->protocol_nodes_key()] = write_nodes_entry(n);
+		r[wanted_node->second->protocol().nodes_key()] = write_nodes_entry(n);
 	}
 }
 
-node::protocol_descriptor const& node::map_protocol_to_descriptor(udp protocol)
+protocol_descriptor::protocol_descriptor(udp protocol)
+	: m_data(protocol == udp::v4() ? v4().m_data : v6().m_data)
+{}
+
+udp protocol_descriptor::protocol() const
 {
-	static protocol_descriptor descriptors[] =
-		{ {udp::v4(), "n4", "nodes"}
-		, {udp::v6(), "n6", "nodes6"} };
+	return m_data.protocol;
+}
 
-	for (int i = 0; i < sizeof(descriptors) / sizeof(protocol_descriptor); ++i)
-	{
-		if (descriptors[i].protocol == protocol)
-			return descriptors[i];
-	}
+char const* protocol_descriptor::family_name() const
+{
+	return m_data.family_name;
+}
 
-	TORRENT_ASSERT_FAIL();
-#ifndef BOOST_NO_EXCEPTIONS
-	throw std::out_of_range("unknown protocol");
-#else
-	std::terminate();
-#endif
+char const* protocol_descriptor::nodes_key() const
+{
+	return m_data.nodes_key;
+}
+
+bool protocol_descriptor::is_v4() const
+{
+	return protocol() == udp::v4();
+}
+
+bool protocol_descriptor::is_v6() const
+{
+	return protocol() == udp::v6();
+}
+
+bool protocol_descriptor::is_native(udp::endpoint const& ep) const
+{
+	return ep.protocol() == protocol();
+}
+
+bool protocol_descriptor::is_native(address const& addr) const
+{
+	return (addr.is_v4() && is_v4())
+		|| (addr.is_v6() && is_v6());
+}
+
+protocol_descriptor protocol_descriptor::v4()
+{
+	static descriptor_data const data = {udp::v4(), "n4", "nodes"};
```

Comment:
for instance, you could have a private constructor, and just do: ``static protocol_descripto ret{udp::v4(), "n4", "nodes"}; return ret;`` and you would also have to return const references wherever you currently return a ``protocol_descriptor``